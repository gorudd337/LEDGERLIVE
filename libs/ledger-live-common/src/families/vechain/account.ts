import type { Account, Operation } from "@ledgerhq/types-live";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";
import { Unit } from "@ledgerhq/types-cryptoassets";

function formatAccountSpecifics(account: Account): string {
  const unit = getAccountUnit(account);
  const formatConfig = {
    disableRounding: true,
    alwaysShowSign: false,
    showCode: true,
  };

  let str = " ";

  str += formatCurrencyUnit(unit, account.spendableBalance, formatConfig) + " spendable. ";

  return str;
}

function formatOperationSpecifics(op: Operation, unit: Unit): string {
  const { additionalField } = op.extra;

  let str = " ";

  const formatConfig = {
    disableRounding: true,
    alwaysShowSign: false,
    showCode: true,
  };

  str +=
    additionalField && !additionalField.isNaN()
      ? `\n    additionalField: ${
          unit ? formatCurrencyUnit(unit, additionalField, formatConfig) : additionalField
        }`
      : "";

  return str;
}

export default {
  formatAccountSpecifics,
  formatOperationSpecifics,
};
