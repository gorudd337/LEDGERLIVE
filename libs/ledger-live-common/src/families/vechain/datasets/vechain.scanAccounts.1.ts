export default [
  {
    name: "vechain seed 1",
    apdus: `
    => e002000015058000002c80000332800000000000000000000000
    <= 410428594655d5b5c95bb2c284ae480caae57df83d5d51e53f3cad0de76c014fdd9c14d47fc40abc126bf41197021d3a2ea6895ce15e50bdf255769b65cd87aaaa7f28383264393834633432664339643439453264344346466463616435393330314166466361374530329000
    => e002000015058000002c80000332800000000000000000000001
    <= 4104356f7b79acc22572d90bf27b7bacec1314c2dbabb2cb23341421b88a8547dca474fe0735b425bbc01e2df377eccd5310d2347bb5b5a122e76783ed1b253adbc028393134323542386335454242383232633166363032373044413342326342393333374638664142389000
    `,
  },
];
