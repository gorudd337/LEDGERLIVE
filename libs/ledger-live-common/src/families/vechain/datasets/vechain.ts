import { AccountRaw } from "@ledgerhq/types-live";
export const vechain1: AccountRaw = {
  id: "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
  seedIdentifier:
    "0428594655d5b5c95bb2c284ae480caae57df83d5d51e53f3cad0de76c014fdd9c14d47fc40abc126bf41197021d3a2ea6895ce15e50bdf255769b65cd87aaaa7f",
  name: "Vechain 1",
  starred: false,
  used: true,
  derivationMode: "vechain",
  index: 0,
  freshAddress: "0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02",
  freshAddressPath: "44'/818'/0'/0/0",
  freshAddresses: [],
  blockHeight: 15156315,
  creationDate: "2023-04-21T13:22:40.000Z",
  operationsCount: 1,
  operations: [],
  pendingOperations: [],
  currencyId: "vechain",
  feesCurrencyId: "vechain/vtho",
  unitMagnitude: 18,
  lastSyncDate: "2023-04-21T14:22:48.395Z",
  balance: "3000000000000000000",
  spendableBalance: "3000000000000000000",
  balanceHistoryCache: {
    HOUR: { balances: [0, 3000000000000000000], latestDate: 1682085600000 },
    DAY: { balances: [0], latestDate: 1682028000000 },
    WEEK: { balances: [0], latestDate: 1681596000000 },
  },
  subAccounts: [
    {
      type: "TokenAccountRaw",
      id: "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho",
      parentId: "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
      starred: false,
      tokenId: "vechain/vtho",
      balance: "3000054150000000000",
      spendableBalance: "3000054150000000000",
      balanceHistoryCache: {
        HOUR: { latestDate: null, balances: [] },
        DAY: { latestDate: null, balances: [] },
        WEEK: { latestDate: null, balances: [] },
      },
      creationDate: "2023-04-21T13:22:40.000Z",
      operationsCount: 1,
      operations: [
        {
          id: "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho-0xa3049ffd1af19c85659e76e50e733098df194ba9df7597e6c654d688c553488b-IN",
          hash: "0xa3049ffd1af19c85659e76e50e733098df194ba9df7597e6c654d688c553488b",
          type: "IN",
          senders: ["0x79028e3d948bd5873ccf58a69089cac105832129"],
          recipients: ["0x82d984c42fc9d49e2d4cffdcad59301affca7e02"],
          accountId:
            "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho",
          blockHash: "0x00e742f48ff7553a1b5276886ca7df27444cafa2086757393ce31ff68422a4b9",
          blockHeight: 15155956,
          extra: {},
          date: "2023-04-21T13:23:00.000Z",
          value: "3000000000000000000",
          fee: "365180000000000000",
        },
      ],
      pendingOperations: [],
      swapHistory: [],
    },
  ],
  swapHistory: [],
};
