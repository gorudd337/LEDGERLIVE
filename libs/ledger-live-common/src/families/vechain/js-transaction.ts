import { Account, TransactionCommon } from "@ledgerhq/types-live";
import BigNumber from "bignumber.js";
import { $Shape } from "utility-types";
import { DEFAULT_GAS_COEFFICIENT, HEX_PREFIX, MAINNET_CHAIN_TAG } from "./constants";
import { Transaction } from "./types";
import { Transaction as ThorTransaction } from "thor-devkit";
import { calculateFee, estimateGas, generateNonce, getBlockRef } from "./utils/transaction-utils";
import { VTHO_ADDRESS } from "./contracts/constants";
import VIP180 from "./contracts/abis/VIP180";
import { calculateTransactionInfo } from "./utils/calculateTransactionInfo";
import { isValid } from "./utils/address-utils";

/**
 * Create an empty VET or VTHO transaction
 *
 * @returns {Transaction}
 */
export const createTransaction = (): Transaction => {
  const chainTag = MAINNET_CHAIN_TAG;

  return {
    family: "vechain",
    body: {
      chainTag,
      // placeholder, if "empty" returns error on send modal open
      blockRef: "0x0000000000000000",
      expiration: 18,
      clauses: [],
      gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
      gas: "0",
      dependsOn: null,
      nonce: generateNonce(),
    },
    amount: BigNumber(0),
    estimatedFees: BigNumber(0),
    recipient: "",
    useAllAmount: false,
    networkInfo: true,
  };
};

/**
 * Apply patch to a transaction
 *
 * @param {Transaction} t
 * @param {TransactionCommon} patch
 * @returns patched transaction
 */
export const updateTransaction = (
  t: Transaction,
  patch: $Shape<TransactionCommon>,
): Transaction => {
  return { ...t, ...patch };
};

/**
 * Prepare transaction before checking status
 *
 * @param {Account} a
 * @param {Transaction} t
 */
export const prepareTransaction = async (
  account: Account,
  transaction: Transaction,
): Promise<Transaction> => {
  const { amount, isTokenAccount } = await calculateTransactionInfo(account, transaction);

  let blockRef = "";

  let clauses: Array<ThorTransaction.Clause> = [];
  if (transaction.recipient && isValid(transaction.recipient)) {
    blockRef = await getBlockRef();
    if (isTokenAccount) {
      clauses = await calculateClausesVtho(transaction, amount);
    } else {
      clauses = await calculateClausesVet(transaction, amount);
    }
  }

  const gas = await estimateGas({
    ...transaction,
    body: { ...transaction.body, clauses: clauses },
  });

  const estimatedFees = await calculateFee(new BigNumber(gas), transaction.body.gasPriceCoef);

  const body = { ...transaction.body, gas, blockRef, clauses };

  return { ...transaction, body, amount, estimatedFees };
};

export const calculateClausesVtho = async (
  transaction: Transaction,
  amount: BigNumber,
): Promise<ThorTransaction.Clause[]> => {
  const clauses: ThorTransaction.Clause[] = [];

  // Get the existing clause or create a blank one
  const updatedClause: ThorTransaction.Clause = {
    to: VTHO_ADDRESS,
    value: 0,
    data: "0x",
  };

  const updatedValues = {
    to: transaction.recipient,
    amount: amount.toFixed(),
  };

  updatedClause.data = VIP180.transfer.encode(updatedValues.to, updatedValues.amount);

  clauses.push(updatedClause);
  return clauses;
};

const calculateClausesVet = async (
  transaction: Transaction,
  amount: BigNumber,
): Promise<ThorTransaction.Clause[]> => {
  const clauses: ThorTransaction.Clause[] = [];

  // Get the existing clause or create a blank one
  const updatedClause: ThorTransaction.Clause = {
    to: null,
    value: 0,
    data: "0x",
  };

  updatedClause.value = `${HEX_PREFIX}${amount.toString(16)}`;
  updatedClause.to = transaction.recipient;

  clauses.push(updatedClause);

  return clauses;
};
