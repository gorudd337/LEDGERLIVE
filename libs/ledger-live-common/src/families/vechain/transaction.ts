import type { Transaction, TransactionRaw, TransactionStatus, TransactionStatusRaw } from "./types";
import {
  formatTransactionStatusCommon as formatTransactionStatus,
  fromTransactionCommonRaw,
  fromTransactionStatusRawCommon,
  toTransactionCommonRaw,
  toTransactionStatusRawCommon,
} from "@ledgerhq/coin-framework/transaction/common";

export const formatTransaction = (t: Transaction): string => {
  return t.estimatedFees.toString();
};

export const fromTransactionRaw = (tr: TransactionRaw): Transaction => {
  const common = fromTransactionCommonRaw(tr);

  return {
    networkInfo: true,
    ...tr,
    ...common,
  };
};

export const toTransactionRaw = (t: Transaction): TransactionRaw => {
  const common = toTransactionCommonRaw(t);

  return {
    ...t,
    ...common,
  };
};

export const fromTransactionStatusRaw = (ts: TransactionStatusRaw): TransactionStatus => {
  const common = fromTransactionStatusRawCommon(ts);

  return {
    ...ts,
    ...common,
  };
};

export const toTransactionStatusRaw = (ts: TransactionStatus): TransactionStatusRaw => {
  const common = toTransactionStatusRawCommon(ts);

  return {
    ...ts,
    ...common,
  };
};

export default {
  formatTransaction,
  formatTransactionStatus,
  fromTransactionRaw,
  toTransactionRaw,
  fromTransactionStatusRaw,
  toTransactionStatusRaw,
};
