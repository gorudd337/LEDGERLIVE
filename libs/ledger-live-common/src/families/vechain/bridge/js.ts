import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";

import { sync, scanAccounts } from "../js-synchronisation";
import { AccountBridge, AccountLike, CurrencyBridge } from "@ledgerhq/types-live";
import { createTransaction, updateTransaction, prepareTransaction } from "../js-transaction";
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import BigNumber from "bignumber.js";
import { calculateMaxFeesToken } from "../utils/calculateTransactionInfo";

const receive: AccountBridge<Transaction>["receive"] = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload: async (): Promise<Record<string, any>> => {
    return {};
  },
  hydrate: (): void => {},
};

const estimateMaxSpendable = async (inputs: {
  account: AccountLike;
  transaction: Transaction;
}): Promise<BigNumber> => {
  const { account, transaction } = inputs;

  if (account.type === "Account") {
    return account.balance;
  }
  if (transaction) {
    const maxTokenFees = await calculateMaxFeesToken();
    const spendable = account.balance.minus(maxTokenFees);
    if (spendable.gt(0)) return account.balance.minus(maxTokenFees);
    return new BigNumber(0);
  } else {
    return account.balance;
  }
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export default { currencyBridge, accountBridge };
