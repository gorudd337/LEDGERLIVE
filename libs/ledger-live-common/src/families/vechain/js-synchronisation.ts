import type { GetAccountShape } from "../../bridge/jsHelpers";
import { BigNumber } from "bignumber.js";
import { makeSync, makeScanAccounts, mergeOps } from "../../bridge/jsHelpers";
import { encodeAccountId } from "../../account";
import eip55 from "eip55";
import { emptyHistoryCache } from "../../account";

import { getAccount, getLastBlock, getOperations, getTokenOperations } from "./api";
import { getTokenById } from "@ledgerhq/cryptoassets/tokens";
// import { Operation } from "@ledgerhq/types-live";
// import { encodeOperationId } from "@ledgerhq/coin-framework/operation";

const getAccountShape: GetAccountShape = async info => {
  const { initialAccount, currency, derivationMode } = info;
  let { address } = info;
  address = eip55.encode(address);

  const oldOperations = initialAccount?.operations || [];
  const startAt = oldOperations.length ? (oldOperations[0].blockHeight || 0) + 1 : 1;

  const accountId = encodeAccountId({
    type: "js",
    version: "2",
    currencyId: currency.id,
    xpubOrAddress: address,
    derivationMode,
  });

  // get the current account balance state depending your api implementation
  const { balance, energy } = await getAccount(address);

  // get the current block height
  const { number: blockHeight } = await getLastBlock();

  // Merge new operations with the previously synced ones
  const newOperations = await getOperations(accountId, address, startAt);

  //Get last token operations
  const vthoAccountId = accountId + "+" + encodeURIComponent("vechain/vtho");
  const VTHOoperations = await getTokenOperations(
    vthoAccountId,
    address,
    "0x0000000000000000000000000000456e65726779",
    1,
  );

  //Generate type "NONE" operations on main account for each token operation to run bot tests
  // VTHOoperations.forEach((operation) => {
  //   const op: Operation = {
  //     id: encodeOperationId(accountId, operation.hash, "NONE" as const),
  //     hash: "",
  //     type: "NONE",
  //     value: new BigNumber(0),
  //     fee: new BigNumber(0),
  //     senders: [],
  //     recipients: [],
  //     blockHash: "",
  //     blockHeight: 0,
  //     accountId,
  //     date: operation.date,
  //     extra: {},
  //   };
  //   newOperations.push(op);
  // });

  const operations = mergeOps(oldOperations, newOperations);

  //Account creation date set to now if there are no operation or at the first operation on the account
  let min_date = -1;
  if (operations.length != 0) {
    const operationsDates = operations.map(c => c.date.getTime());
    operationsDates.concat(VTHOoperations.map(c => c.date.getTime()));
    min_date = Math.min(...operationsDates);
  }

  const shape = {
    ...info,
    id: accountId,
    balance: new BigNumber(balance),
    creationDate: min_date != -1 ? new Date(min_date) : new Date(),
    spendableBalance: new BigNumber(balance),
    operationsCount: operations.length,
    operations: operations,
    blockHeight,
    feesCurrency: getTokenById("vechain/vtho"),
    subAccounts: [
      {
        type: "TokenAccount" as const,
        id: vthoAccountId,
        parentId: accountId,
        token: getTokenById("vechain/vtho"),
        balance: new BigNumber(energy),
        spendableBalance: new BigNumber(energy),
        creationDate: min_date != -1 ? new Date(min_date) : new Date(),
        operationsCount: VTHOoperations.length,
        operations: VTHOoperations,
        blockHeight,
        pendingOperations:
          (initialAccount?.subAccounts && initialAccount.subAccounts[0]?.pendingOperations) || [],
        starred: (initialAccount?.subAccounts && initialAccount.subAccounts[0]?.starred) || false,
        balanceHistoryCache:
          (initialAccount?.subAccounts && initialAccount.subAccounts[0]?.balanceHistoryCache) ||
          emptyHistoryCache,
        swapHistory: [],
      },
    ],
  };

  return { ...shape, operations };
};

export const scanAccounts = makeScanAccounts({ getAccountShape });
export const sync = makeSync({ getAccountShape });
