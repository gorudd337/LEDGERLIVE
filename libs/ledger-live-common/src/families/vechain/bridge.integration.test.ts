import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import type { AccountRaw, DatasetTest } from "@ledgerhq/types-live";
import type { Transaction } from "./types";
import { fromTransactionRaw } from "./transaction";
import { DEFAULT_GAS_COEFFICIENT, MAINNET_CHAIN_TAG } from "./constants";
import { vechain1 } from "./datasets/vechain";
import { generateNonce } from "./utils/transaction-utils";
import BigNumber from "bignumber.js";
import {
  setSupportedCurrencies,
  listSupportedCurrencies,
  getCryptoCurrencyById,
} from "../../currencies";
import vechainScanAccounts1 from "./datasets/vechain.scanAccounts.1";
import { AmountRequired, NotEnoughBalance } from "@ledgerhq/errors";
import VIP180 from "./contracts/abis/VIP180";
import { CryptoCurrencyId } from "@ledgerhq/types-cryptoassets";

const listSupported = listSupportedCurrencies();
listSupported.push(getCryptoCurrencyById("vechain"));
setSupportedCurrencies(listSupported.map(c => c.id) as CryptoCurrencyId[]);

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    vechain: {
      FIXME_ignoreAccountFields: ["balance", "spendableBalance", "estimateMaxSpendable"],
      scanAccounts: vechainScanAccounts1,
      accounts: [
        {
          raw: vechain1 as AccountRaw,
          implementations: ["js"],
          transactions: [
            {
              name: "Send VET",
              transaction: fromTransactionRaw({
                family: "vechain",
                estimatedFees: new BigNumber(0),
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "1000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                      value: "1000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("1000000000000000000"),
                estimatedFees: new BigNumber("210000000000000000"),
                totalSpent: new BigNumber("1000000000000000000"),
                errors: {},
                warnings: {},
              },
            },
            {
              name: "Send VTHO",
              transaction: fromTransactionRaw({
                family: "vechain",
                subAccountId:
                  "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho",
                estimatedFees: new BigNumber("210000000000000000"),
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "1000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x0000000000000000000000000000456e65726779",
                      value: 0,
                      data: VIP180.transfer.encode(
                        "0x17733CAb76d9A2112576443F21735789733B1ca3",
                        "9000000000000000000",
                      ),
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("1000000000000000000"),
                estimatedFees: new BigNumber("665180000000000000"),
                totalSpent: new BigNumber("1665180000000000000"),
                errors: {},
                warnings: {},
              },
            },
            {
              name: "Amount required",
              transaction: fromTransactionRaw({
                family: "vechain",
                estimatedFees: new BigNumber(0),
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [{ to: "", value: 0, data: "0x" }],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                errors: {
                  amount: new AmountRequired(),
                },
                warnings: {},
              },
            },
            {
              name: "VET balance not enough",
              transaction: fromTransactionRaw({
                family: "vechain",
                estimatedFees: new BigNumber(0),
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "20000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                      value: "20000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("20000000000000000000"),
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
                totalSpent: new BigNumber("0"),
                estimatedFees: new BigNumber("0"),
              },
            },
            {
              name: "VTHO balance not enough",
              transaction: fromTransactionRaw({
                family: "vechain",
                subAccountId:
                  "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho",
                estimatedFees: new BigNumber(0),
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "20000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                      value: "20000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("20000000000000000000"),
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
                totalSpent: new BigNumber("0"),
                estimatedFees: new BigNumber("0"),
              },
            },
          ],
          FIXME_tests: [
            "balance is sum of ops", //the balance depends on VTHO and it's earned without operations
            "empty transaction is equals to itself", //nonce is not deterministic
            "ref stability on self transaction", //blockref is not deterministic
            "can be run in parallel and all yield same results", //blockref is not deterministic
          ],
        },
      ],
    },
  },
};

testBridge(dataset);
