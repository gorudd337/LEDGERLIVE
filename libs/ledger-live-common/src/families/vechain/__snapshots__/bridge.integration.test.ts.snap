// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vechain currency bridge scanAccounts vechain seed 1 1`] = `
Array [
  Object {
    "currencyId": "vechain",
    "derivationMode": "vechain",
    "feesCurrencyId": "vechain/vtho",
    "freshAddress": "0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02",
    "freshAddressPath": "44'/818'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02",
        "derivationPath": "44'/818'/0'/0/0",
      },
    ],
    "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
    "index": 0,
    "name": "Vechain 1",
    "nfts": undefined,
    "operationsCount": 1,
    "pendingOperations": Array [],
    "seedIdentifier": "0428594655d5b5c95bb2c284ae480caae57df83d5d51e53f3cad0de76c014fdd9c14d47fc40abc126bf41197021d3a2ea6895ce15e50bdf255769b65cd87aaaa7f",
    "starred": false,
    "subAccounts": Array [],
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": true,
  },
  Object {
    "approvals": undefined,
    "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho",
    "operationsCount": 1,
    "parentId": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
    "pendingOperations": Array [],
    "starred": false,
    "swapHistory": Array [],
    "tokenId": "vechain/vtho",
    "type": "TokenAccountRaw",
  },
  Object {
    "currencyId": "vechain",
    "derivationMode": "vechain",
    "feesCurrencyId": "vechain/vtho",
    "freshAddress": "0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8",
    "freshAddressPath": "44'/818'/0'/0/1",
    "freshAddresses": Array [
      Object {
        "address": "0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8",
        "derivationPath": "44'/818'/0'/0/1",
      },
    ],
    "id": "js:2:vechain:0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8:vechain",
    "index": 1,
    "name": "Vechain 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "0428594655d5b5c95bb2c284ae480caae57df83d5d51e53f3cad0de76c014fdd9c14d47fc40abc126bf41197021d3a2ea6895ce15e50bdf255769b65cd87aaaa7f",
    "starred": false,
    "subAccounts": Array [],
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": false,
  },
  Object {
    "approvals": undefined,
    "id": "js:2:vechain:0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8:vechain+vechain%2Fvtho",
    "operationsCount": 0,
    "parentId": "js:2:vechain:0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8:vechain",
    "pendingOperations": Array [],
    "starred": false,
    "swapHistory": Array [],
    "tokenId": "vechain/vtho",
    "type": "TokenAccountRaw",
  },
]
`;

exports[`vechain currency bridge scanAccounts vechain seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
      "blockHash": "0x00e742f27e0dddaa6df6fc23fb39a24acb50d57dbbe114f64fb06ddfa4d358e8",
      "blockHeight": 15155954,
      "contract": undefined,
      "extra": Object {},
      "fee": "210000000000000000",
      "hash": "0x901109642a6f56495bdd652cd7c9fb96ee476377b976551b46a05437567af58c",
      "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain-0x901109642a6f56495bdd652cd7c9fb96ee476377b976551b46a05437567af58c-IN",
      "operator": undefined,
      "recipients": Array [
        "0x82d984c42fc9d49e2d4cffdcad59301affca7e02",
      ],
      "senders": Array [
        "0x79028e3d948bd5873ccf58a69089cac105832129",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "3000000000000000000",
    },
  ],
  Array [
    Object {
      "accountId": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho",
      "blockHash": "0x00e742f48ff7553a1b5276886ca7df27444cafa2086757393ce31ff68422a4b9",
      "blockHeight": 15155956,
      "contract": undefined,
      "extra": Object {},
      "fee": "365180000000000000",
      "hash": "0xa3049ffd1af19c85659e76e50e733098df194ba9df7597e6c654d688c553488b",
      "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho-0xa3049ffd1af19c85659e76e50e733098df194ba9df7597e6c654d688c553488b-IN",
      "operator": undefined,
      "recipients": Array [
        "0x82d984c42fc9d49e2d4cffdcad59301affca7e02",
      ],
      "senders": Array [
        "0x79028e3d948bd5873ccf58a69089cac105832129",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "3000000000000000000",
    },
  ],
  Array [],
  Array [],
]
`;
